### PR TITLE
db4: fix pthread configure detection

### PR DESCRIPTION
**Pull request by paparodeo**
## Description of changes

add header stdlib.h so pthread detection doesn't fail do to a missing function prototype for exit.
    
    configure:21863: clang -o conftest -O3   -D_GNU_SOURCE -D_REENTRANT
    conftest.c  >&5
    conftest.c:52:2: error: call to undeclared library function 'exit' with
    type 'void (int) __attribute__((noreturn))'; ISO C99 and later do not
    support implicit function declarations [-Wimplicit-function-declaration]
       52 |         exit (
          |         ^
    conftest.c:52:2: note: include the header <stdlib.h> or explicitly
    provide a declaration for 'exit'

additionally, add a leading space to line 218 to uncorrupt the patch -- git apply was complaining and failed to apply without fix.

this reverts the previous workaround https://github.com/NixOS/nixpkgs/pull/320187

testing: `nix-build -E 'with import ./.{}; db4.override { stdenv = clangStdenv; }` builds

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by paparodeo** [comment]
cc: @@drupol @RossComputerGuy

**Comment by drupol** [comment]
Don't you think this patch should be submitted upstream instead ?

**Comment by paparodeo** [comment]
> Don't you think this patch should be submitted upstream instead ?

it is fixed for db5, tho there are still a bunch of patches in nix for that version as well, one being that main needs to return an int. the patches for the various db versions are:
https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/libraries/db

and if i untar the source for db4 the dates are all `Apr 12 2010` and the homepage url points to https://www.oracle.com/database/technologies/related/berkeleydb.html so not sure where to send a patch to.

so i get the feeling that this version is pretty old so am not sure where to upstream the patch.

**Comment by drupol** [comment]
Fair enough.

File: pkgs/development/libraries/db/generic.nix [hunk 1]
```diff
@@ -67,8 +67,7 @@ stdenv.mkDerivation (rec {
       (if compat185 then "--enable-compat185" else "--disable-compat185")
     ]
     ++ lib.optional dbmSupport "--enable-dbm"
-    ++ lib.optional stdenv.isFreeBSD "--with-pic"
-    ++ (drvArgs.configureFlags or []);
+    ++ lib.optional stdenv.isFreeBSD "--with-pic";
```
Comment (drupol) [inline pkgs/development/libraries/db/generic.nix:70]:
```suggestion
    ++ lib.optionals stdenv.isFreeBSD [ "--with-pic" ];
```
Opinionated: I would use `lib.optionals` instead, since `configureFlags` is a list. Semantically, it makes more sense to me.

File: pkgs/development/libraries/db/clang-4.8.patch [hunk 2]
```diff
@@ -183,7 +203,7 @@ diff -ur a/dist/aclocal/sequence.m4 b/dist/aclocal/sequence.m4
  			$db_cv_seq_type l;
  			unsigned $db_cv_seq_type u;
 @@ -59,7 +62,9 @@
-				return (1);
```
[the remark this answers is not in the document]
Comment (paparodeo) [reply pkgs/development/libraries/db/clang-4.8.patch:186]: so i added a space to un-corrupt the diff. i guess generating a diff the first character is a space, + or -, except this line was missing the space and `git apply` as complaining about the patch being corrupted so i fixed it. i can revert it as patch doesn't seem to care.

[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/development/libraries/db/clang-4.8.patch:186]: OK don't, I was just curious.

File: pkgs/development/libraries/db/generic.nix (hunk 1 above)
Comment (paparodeo) [reply pkgs/development/libraries/db/generic.nix:70]: so this change is just a revert of the previous workaround for pkgsLLVM so i'd prefer to leave as is (it is the first commit). my change is the 2nd commit which modifies the patch. in general, i do try to prefer optionals over optional.

Comment (drupol) [reply pkgs/development/libraries/db/generic.nix:70]: It's totally fine and optional, no worries.

**Comment by paparodeo** [comment]
FWIW this is also needed to fix compiling with gcc14.

```
nix-build -E 'with import ./.{}; db4.override { stdenv = gcc14Stdenv; }'
```

fails the same way as if using `clangStdenv`

File: pkgs/development/libraries/db/clang-4.8.patch
Comment (emilazy) [inline pkgs/development/libraries/db/clang-4.8.patch:1]: Maybe it would be better to split this out into a separate patch? Or at least rename this to something more generic, if it’s also tackling a modern LLVM thing and recent GCC?

Comment (paparodeo) [reply pkgs/development/libraries/db/clang-4.8.patch:1]: it is the db 4.8 clang patch which fixes similar issues in the same file. eg: main not returning an int. i wouldn't be surprised if new gcc needs all the same fixes in this file too. and the naming matches the other autoconf patches for the db 5.3 and db 6.0.

i'm not quite sure what your asking for -- a re-organization of the package to rename the patches and test them all on new versions of gcc to split out the intersection of fixes needed by both gcc and clang to the patch can be named appropriately?

thanks for taking a look.

Comment (emilazy) [reply pkgs/development/libraries/db/clang-4.8.patch:1]: Ah, sorry, I misinterpreted what the 4.8 was referring to. Something like `db-4.8-modern-compilers` or your `autoconf-modernization` suggestion or whatever seem fine to me.

**Comment by paparodeo** [comment]
change seems fine to me as is and i don't want to put any more time into it so closing.

**Comment by emilazy** [comment]
I’m happy to merge this as‐is if you wanted to reopen. Sorry for nitpicking about the filename when I lacked the full context of the patches.

**Review by emilazy — APPROVED** [review]
Successfully built `db4` on `x86_64-darwin` after rebasing this on `master` locally.

**Comment by vobloeb** [comment]
`"--with-mutex=POSIX/pthreads";` used to fix GCC-14, but this PR fixes only Clang, so GCC-14 is broken again
